cmd/snapd-generator/main.c

```diff
@@ -30,19 +31,168 @@
+			      const char *what,
+			      const char *where)
+{
+	// Replace directory separators with dashes to build the unit name.
+	// WARNING this is limited to directories that do no contain dashes,
+	// which is fine at the moment as our mountpoints are
+	// /usr/lib/{modules,firmware}.
+	char *unit_name SC_CLEANUP(sc_cleanup_string) = NULL;
+	// (... + 1) to remove the initial '/'
+	unit_name = sc_strdup(where + 1);
+	for (char *p = unit_name; (p = strchr(p, '/')) != NULL; *p = '-');
+
+	// Construct the file name for a new systemd mount unit.
+	char fname[PATH_MAX + 1] = { 0 };
+	sc_must_snprintf(fname, sizeof fname,
+			 "%s/%s.mount",	 normal_dir, unit_name);
+
+	// Open the mount unit and write the contents.
+	FILE *f SC_CLEANUP(sc_cleanup_file) = NULL;
+	f = fopen(fname, "w");
+	if (!f) {
+		fprintf(stderr, "cannot write to %s: %m\n", fname);
+		return 1;
+	}
+	fprintf(f, "[Unit]\n");
+	fprintf(f, "Description=Early mount of kernel drivers tree\n");
+	fprintf(f, "DefaultDependencies=no\n");
+	fprintf(f, "After=systemd-remount-fs.service\n");
+	fprintf(f, "Before=sysinit.target\n");
+	fprintf(f, "Before=systemd-udevd.service systemd-modules-load.service\n");
+	fprintf(f, "Before=umount.target\n");
+	fprintf(f, "Conflicts=umount.target\n");
+	fprintf(f, "\n");
+	fprintf(f, "[Mount]\n");
+	fprintf(f, "What=%s\n", what);
+	fprintf(f, "Where=%s\n", where);
+	fprintf(f, "Options=bind,shared\n");
+
+	// Wanted by sysinit.target.wants - create folders if needed and symlink
+
+	char wants_d[PATH_MAX + 1] = { 0 };
+	sc_must_snprintf(wants_d, sizeof wants_d,
+			 "%s/sysinit.target.wants", normal_dir);
+	if (mkdir(wants_d, 0755) != 0 && errno != EEXIST) {
+		fprintf(stderr, "cannot create %s directory: %m\n", wants_d);
+		return 1;
+	}
+
+	char target[PATH_MAX + 1] = { 0 };
+	char lnpath[PATH_MAX + 1] = { 0 };
+	sc_must_snprintf(target, sizeof target, "../%s.mount", unit_name);
+	sc_must_snprintf(lnpath, sizeof lnpath,
+			 "%s/%s.mount", wants_d, unit_name);
+	if (symlink(target, lnpath) != 0) {
+		fprintf(stderr, "cannot create symlink %s: %m\n", lnpath);
+		return 1;
+	}
+
+	return 0;
+}
+
+#define MAJOR_LOOP_DEV 7
+#define SNAPD_DRIVERS_TREE_DIR "/var/lib/snapd/kernel"
+
+static int ensure_kernel_drivers_mounts(const char *normal_dir)
+{
+	// Find active kernel name and revision by looking at what was
+	// mounted in /run/mnt/kernel by snap-bootstrap.
+
+	const char* kern_mnt_dir = "/run/mnt/kernel";
+	// Find mount information about kern_mnt_dir
+	sc_mountinfo *mounts SC_CLEANUP(sc_cleanup_mountinfo) = NULL;
+	mounts = sc_parse_mountinfo("/proc/1/mountinfo");
+	if (!mounts) {
+		fprintf(stderr, "cannot open or parse /proc/1/mountinfo\n");
+		return 1;
+	}
+	sc_mountinfo_entry *kern_minfo = find_dir_mountinfo(mounts, kern_mnt_dir);
+	if (!kern_minfo) {
+		// This is not Ubuntu Core / hybrid, do nothing and do not fail
+		return 0;
+	}
+
+	// Mount source should be a snap
+	if (!sc_streq(kern_minfo->fs_type, "squashfs")) {
```

Comment:
I mussed this earlier.


I think the mount source could also be a FUSE, right?

```
6756 7355 0:131 / /snap/core20/2105 ro,nodev,relatime shared:1127 - fuse.snapfuse snapfuse ro,user_id=0,group_id=0,allow_other
```

If this happens we should just not do things cleanly, without causing things to fail on startup.

Reply:
But then it would be a container and we would have detected that `/run/mnt/kernel` is not a mount point as we cannot have a kernel snap around, even less an initramfs creating such mount point. And it does not look like `/run/mnt` leaks to the container mount namespace even when run on UC.